modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1584,17 +1584,55 @@ void GDScriptAnalyzer::resolve_assignable(GDScriptParser::AssignableNode *p_assi
-				if (is_variable && is_type_compatible(initializer_type, specified_type, true, p_assignable->initializer)) {
-					mark_node_unsafe(p_assignable->initializer);
-					static_cast<GDScriptParser::VariableNode *>(p_assignable)->use_conversion_assign = true;
+			} else {
+				Variant::Type specified_builtin_type = type_from_metatype(specified_type).builtin_type;
+				Variant::Type initializer_builtin_type = type_from_metatype(initializer_type).builtin_type;
+
+				enum Compatible {
+					SAFE,
+					UNSAFE,
+					ERROR,
+				};
+				Compatible compatible;
+
+				if (is_type_compatible(specified_type, initializer_type, true, p_assignable->initializer)) {
+					compatible = SAFE;
+				} else if (p_assignable->initializer->is_constant) {
+					compatible = ERROR;
+				} else if (specified_builtin_type == Variant::OBJECT && initializer_builtin_type == Variant::OBJECT && is_type_compatible(initializer_type, specified_type, true)) {
+					// For example: `var label: Label = $Node`. It's unsafe, but allowed.
+					compatible = UNSAFE;
 				} else {
-					push_error(vformat(R"(Cannot assign a value of type %s to %s "%s" with specified type %s.)", initializer_type.to_string(), p_kind, p_assignable->identifier->name, specified_type.to_string()), p_assignable->initializer);
+					compatible = ERROR;
 				}
+
+				if (compatible == ERROR) {
+					push_error(vformat(R"(Cannot assign a value of type %s to %s "%s" with specified type %s.)", initializer_type.to_string(), p_kind, p_assignable->identifier->name, specified_type.to_string()), p_assignable->initializer);
+				} else {
 #ifdef DEBUG_ENABLED
-			} else if (specified_type.builtin_type == Variant::INT && initializer_type.builtin_type == Variant::FLOAT) {
-				parser->push_warning(p_assignable->initializer, GDScriptWarning::NARROWING_CONVERSION);
+					if (specified_builtin_type == Variant::INT && initializer_builtin_type == Variant::FLOAT) {
+						parser->push_warning(p_assignable->initializer, GDScriptWarning::NARROWING_CONVERSION);
+					}
 #endif
+
+					if (p_assignable->initializer->is_constant && initializer_builtin_type != specified_builtin_type) {
+						Variant converted;
+						const Variant *value = &p_assignable->initializer->reduced_value;
+						Callable::CallError ce;
+						Variant::construct(specified_builtin_type, converted, &value, 1, ce);
+						if (ce.error == Callable::CallError::CALL_OK) {
+							p_assignable->initializer->reduced_value = converted;
+						} else {
+							// Normally, this should not happen.
+							ERR_PRINT(vformat(R"(Cannot convert %s "%s" initializer value of type %s to the specified type %s.)", p_kind, p_assignable->identifier->name, initializer_type.to_string(), specified_type.to_string()));
+						}
+					} else if (compatible == UNSAFE) {
+						mark_node_unsafe(p_assignable->initializer);
+						if (is_variable) {
+							static_cast<GDScriptParser::VariableNode *>(p_assignable)->use_conversion_assign = true;
+						}
```

Comment:
This fragment is repeated above, but I think it's better to leave duplication here.